pkg/unstruct/v1alpha2/kubernetes.go

```diff
@@ -17,50 +17,123 @@ limitations under the License.
 package v1alpha2
 
 import (
-	"context"
 	"strings"
 
-	kclient "github.com/openebs/maya/pkg/kubernetes/client/v1alpha2"
-	provider "github.com/openebs/maya/pkg/provider/v1alpha1"
+	k8s "github.com/openebs/maya/pkg/client/k8s/v1alpha1"
 	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
-	"sigs.k8s.io/controller-runtime/pkg/client"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/dynamic"
 )
 
-// kubeclient enables kubernetes API operations on catalog
+// getClientsetFn is a typed function that
+// abstracts fetching of clientset
+type getClientsetFn func() (clientset dynamic.Interface, err error)
+
+// CreateFn is a typed function that abstracts
+// creating of unstructured object
+type CreateFn func(
+	cli dynamic.Interface,
+	obj *unstructured.Unstructured,
+	opts *CreateOption,
+) (*unstructured.Unstructured, error)
+
+// GetFn is a typed function that abstracts
+// fetching of unstructured object
+type GetFn func(
+	cli dynamic.Interface,
+	name string,
+	namespace string, opt *GetOption,
+) (*unstructured.Unstructured, error)
+
+// DeleteFn is a typed function that abstract  deletion
+// of unstructured object
+type DeleteFn func(
+	cli dynamic.Interface,
+	obj *unstructured.Unstructured,
+	opt *DeleteOption,
+) error
+
+// Kubeclient enables kubernetes API operations on catalog
 // instance
-type kubeclient struct {
-	client.Client  // kubernetes client
-	kclient.Handle // manage kubernetes client & enable mocking
+type Kubeclient struct {
+	// clientset refers to clientset
+	// that will be responsible to
+	// make kubernetes API calls
+	clientset dynamic.Interface
+
+	// functions useful during mocking
+	getClientset getClientsetFn
+	create       CreateFn
+	get          GetFn
+	delete       DeleteFn
 }
 
-// kubeclientBuildOption defines the abstraction to build
-// a kubeclient instance
-type kubeclientBuildOption func(*kubeclient)
+// KubeclientBuildOption defines the abstraction to build
+// a Kubeclient instance
+type KubeclientBuildOption func(*Kubeclient)
 
-func withDefaults(k *kubeclient) error {
-	if k.Client == nil {
-		cli, err := kclient.New()
+// withDefaults sets default options for Kubeclient
+func withDefaults(k *Kubeclient) error {
+	if k.clientset == nil {
```

Comment:
// TODO // Next PR
This need not be set from defaults. All the function properties should be set.
clientset will be set at `getClientsetOrCached()`. 
`getClientsetOrCached()` in turn will make use of function properties.